Mathlib/NumberTheory/Padics/AddChar.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2025 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.NumberTheory.Padics.MahlerBasis
+
+/-!
+# Additive characters of `ℤ_[p]`
+
+We show that for any normed `ℤ_[p]`-algebra `R`, there is a canonical bijection between continuous
+additive characters `ℤ_[p] → R` and topologically nilpotent elements of `R`, given by sending `κ`
+to the element `κ 1 - 1`. This is used to define the Mahler transform for `p`-adic measures.
+
+Note that if the norm on `R` is not strictly multiplicative, then the condition that `κ 1 - 1` be
+topologically nilpotent is strictly weaker than assuming `‖κ 1 - 1‖ < 1`, although they are of
+course equivalent if `NormMulClass R` holds.
+-/
+
+open scoped fwdDiff
+open Filter Topology
+
+lemma fwdDiff_addChar_eq {M R : Type*} [AddCommMonoid M] [Ring R]
+    (φ : AddChar M R) (x h : M) (n : ℕ) : Δ_[h]^[n] φ x = (φ h - 1) ^ n * φ x := by
+  induction n generalizing x with
+  | zero => simp
+  | succ n IH =>
+    simp only [pow_succ, Function.iterate_succ_apply', fwdDiff, IH, ← mul_sub, mul_assoc]
+    rw [sub_mul, ← AddChar.map_add_eq_mul, add_comm h x, one_mul]
+
+variable {p : ℕ} [Fact p.Prime]
+
+/-- If two continuous additive characters of `ℤ_[p]` agree at 1, they agree everwhere. -/
+lemma AddChar.eq_of_apply_one_eq {M : Type*} [Monoid M] [TopologicalSpace M] [T2Space M]
+    {κ₁ κ₂ : AddChar ℤ_[p] M} (hκ₁ : Continuous κ₁) (hκ₂ : Continuous κ₂) (h : κ₁ 1 = κ₂ 1) :
+    κ₁ = κ₂ := by
+  refine DFunLike.coe_injective <| PadicInt.denseRange_natCast.equalizer hκ₁ hκ₂ (funext fun n ↦ ?_)
+  simp only [Function.comp_apply, ← nsmul_one, h, AddChar.map_nsmul_eq_pow]
+
+variable {R : Type*} [NormedRing R] [Algebra ℤ_[p] R] [IsBoundedSMul ℤ_[p] R]
+  [IsUltrametricDist R]
+
+lemma AddChar.tendsto_apply_one_sub_pow {κ : AddChar ℤ_[p] R} (hκ : Continuous κ) :
+    Tendsto (fun n ↦ (κ 1 - 1) ^ n) atTop (𝓝 0) := by
+  refine (PadicInt.fwdDiff_tendsto_zero ⟨κ, hκ⟩).congr fun n ↦ ?_
+  simpa only [AddChar.map_zero_eq_one, mul_one] using fwdDiff_addChar_eq κ 0 1 n
+
+namespace PadicInt
+variable [CompleteSpace R]
+
+/-- The unique continuous additive character of `ℤ_[p]` mapping `1` to `1 + r`. -/
```

Comment:
Given your lemma above, it shuold be very easy (and useful, IMHO) it add a lemma saying that if `addChar_of_value_at_one r = addChar_of_value_at_one s` then `r=s`, hence fully justifying this "uniqueness" statement. Also, why is `r` below explicit while being implicit almost everywhere else?

Reply:
As for `r` being explicit: I learned from Eric Wieser that the rule "all arguments inferrable from another argument should be implicit" only applies to lemmas, not to defs, because otherwise we get objects which print incomprehensibly in the tactic state. If we made `r` implicit here, then we'd get tactic state output like `addChar_of_value_at_one ⋯` where `⋯` is a placeholder for a proof that some (invisible!) `r` has some property.

Reply:
I'm not sure what you mean about "justifying this uniqueness" here: the uniqueness that I am claiming in the docstring is that `κ` is uniquely determined by `r`, not that `r` is uniquely determined by `κ` (which is trivial). I think all the uniqueness statements one can sensibly ask for in this setting are bundled together in the fact that `continuousAddCharEquiv` is a bijection, so I'm wary of adding more as separate lemmas.

Reply:
Fine, I agree.